u8timeseries/backtesting/backtesting.py

```diff
@@ -259,3 +252,117 @@ def plot_residuals_analysis(residuals: TimeSeries, num_bins: int = 20):
+        respective hyperparameter.
+    train_series
+        The TimeSeries instance used for training (and also validation in split mode).
+    test_series
+        The TimeSeries instance used for validation in split mode.
+    fcast_horizon_n
+        The integer value of the forecasting horizon used in expanding window mode.
+    num_predictions:
+        The number of train/prediction cycles performed in one iteration of expanding window mode.
+    metric:
+        The function name (as string) of a metrics function from the metrics module used for evaluating performance.
+    verbose:
+        Whether to print progress.
+
+    Returns
+    -------
+    ForecastingModel
+        A 'model_class' instance with the best-performing hyperparameters from the given selection.
+    """
+
+    raise_if_not(hasattr(metrics, metric), "'metric' must be an attribute of u8timeseries.metrics.", logger)
+
+    backtest_start_time = train_series.end_time() - (num_predictions + fcast_horizon_n) * train_series.freq()
+    metric_function = getattr(metrics, metric)
+
+    min_error = float('inf')
+    best_param_combination = {}
+    params_cross_product = list(product(*parameters.values()))
+    iterator = _build_tqdm_iterator(params_cross_product, verbose)
+
+    # iterate through all combinations of the provided parameters and choose the best one
+    for param_combination in iterator:
+        param_combination_dict = dict(list(zip(parameters.keys(), param_combination)))
+        model = model_class(**param_combination_dict)
+        if (val_series is None):  # expanding window mode
+            backtest_forecast = backtest_forecasting(train_series, model, backtest_start_time, fcast_horizon_n)
+            error = metric_function(backtest_forecast, train_series)
+        else:  # split mode
+            model.fit(train_series)
+            error = metric_function(model.predict(len(val_series)), val_series)
+        if (error < min_error):
+            min_error = error
+            best_param_combination = param_combination_dict
+    logger.info('Chosen parameters: ' + str(best_param_combination))
+    return model_class(**best_param_combination)
+
+
+def explore_models(train_series: TimeSeries, val_series: TimeSeries):
+
+    model_parameter_tuples = [
+        (AutoARIMA, {}),
+        (FFT, {
+            'nr_freqs_to_keep': [5, 10, 25, 50, 100],
+            'trend': [None, 'poly', 'exp']
+        }),
+        (Theta, {
+            'theta': list(range(3, 10))
+        }),
+        (Prophet, {})
+    ]
+
+    for model_class, params in model_parameter_tuples:
+
+        if (len(params.keys()) > 0):
+            model = backtest_gridsearch(model_class, params, train_series, val_series)
+        else:
+            model = model_class()
+
+        # fit and predict
+        model.fit(train_series)
+        predictions = model.predict(len(val_series))
+
+        # visualize
+        print(str(model), ", MAPE:", metrics.mape(predictions, val_series))
```

Comment:
I think the metric function could be an (optional) parameter of the function `explore_models()`

Reply:
done